### ISSUE TITLE
Improve scatter operations speed on GPU

### ISSUE THREAD

**Issue by TristanBilot** [ml-explore/mlx#506]
**Describe the bug**
We benchmarked the `scatter_add` and `scatter_max` operations as part of our developments on [mlx-graphs](https://github.com/TristanBilot/mlx-graphs). 

We used as baseline for comparison the efficient scatter operations from [torch_scatter](https://github.com/rusty1s/pytorch_scatter) on CPU and MPS backends.

Here are the results in milliseconds, averaged on 5 experiments. nan values refer to non-implemented mps operation.
Config: **M1 Pro (2E+8P+16GPU)**

| Operation                                                     | mlx_gpu | mlx_cpu | torch_mps | torch_cpu | mlx_gpu/torch_cpu speedup | mlx_gpu/torch_mps speedup |
|---------------------------------------------------------------|-------|-------|-------|-------|-----------------------|-----------------------|
| scatter_add / indices=100000, dst_shape=(100, 64), src_shape=(100000, 64) |   6.18 |  22.94 | 146.09 |   7.52 |    +21% |  +2264% |
| scatter_add / indices=1000000, dst_shape=(10, 64), src_shape=(1000000, 64) |  64.14 | 228.08 | 18766.79 |  83.65 |    +30% | +29157% |
| scatter_add / indices=10000, dst_shape=(1000, 64), src_shape=(10000, 64) |   2.37 |   2.32 |   3.17 |   1.56 |    -34% |    +33% |
| scatter_max / indices=100000, dst_shape=(100, 64), src_shape=(100000, 64) |   6.07 |  22.86 | nan |  18.42 |   +203% | nan |
| scatter_max / indices=1000000, dst_shape=(10, 64), src_shape=(1000000, 64) |  62.83 | 228.12 | nan | 182.18 |   +189% | nan |
| scatter_max / indices=10000, dst_shape=(1000, 64), src_shape=(10000, 64) |   2.08 |   2.35 | nan |   5.11 |   +145% | nan |

**To Reproduce**
Install the benchmark branch of mlx-graphs and create a new env:
```shell
git clone https://github.com/TristanBilot/mlx-graphs.git -b benchmark
cd mlx-graphs
CONDA_SUBDIR=osx-arm64 conda create -n mlx_graphs python=3.10
conda activate mlx_graphs
```

Install mlx-graphs locally and install dependencies:
```shell
pip install -e . 
pip install mlx==0.0.10
pip install torch_geometric torch_scatter torch
```

Run the benchmark:
```shell
cd benchmark
python run_benchmark.py
```

**Expected behavior**
As shown in the previous table, we only notice ~+30% speedup improvements between mlx with gpu and torch's scatter implementation that runs on cpu, and even ~-30% when the number of unique destinations is 1000. In these experiments, `src_shape` refers to the shape of the values to scatter, and `dst_shape` is an all-zero output array. 
Do you think there are any ways to improve the runtime of these scattering operations on gpu? 
Maybe it has something to do with atomic operations? @awni 

**Desktop:**
 - OS Version: MacOS 14.2

**Comment by awni** [comment]
Thanks for the benchmarks! We will take a look. If this has to due with collisions in the atomic output as suspected there should be a lot of room for improvement

**Comment by TristanBilot** [comment]
Hi, here are some updates:
The initial results I shared in the previous table were obtained after running a preliminary scatter operation before the actual measured scatter. Indeed, I've observed that the initial run of `scatter_*` operations is significantly slower than subsequent executions. To address this, I executed the same scatter operation twice and only measured the runtime of the second run to avoid the 'cold start' delay experienced in the first run. However, this approach isn't practical in real-world scenarios, so here are the benchmark results from running the scatter operation for the first time, as it would typically be done in actual use cases.

| Operation                                                     | mlx_gpu | mlx_cpu | torch_mps | torch_cpu | mlx_gpu/torch_cpu speedup | mlx_gpu/torch_mps speedup |
|---------------------------------------------------------------|-------|-------|-------|-------|-----------------------|-----------------------|
| scatter_add / indices=100000, dst_shape=(100, 64), src_shape=(100000, 64) |  35.20 |  28.71 | 124.97 |  14.04 |    -60% |   +254% |
| scatter_add / indices=1000000, dst_shape=(10, 64), src_shape=(1000000, 64) | 154.26 | 292.33 | 154.38 |  88.58 |    -42% |     +0% |
| scatter_add / indices=10000, dst_shape=(1000, 64), src_shape=(10000, 64) |  14.87 |   3.82 |  56.71 |   1.75 |    -88% |   +281% |
| scatter_max / indices=100000, dst_shape=(100, 64), src_shape=(100000, 64) |  35.23 |  28.53 |  99.75 |  13.44 |    -61% |   +183% |
| scatter_max / indices=1000000, dst_shape=(10, 64), src_shape=(1000000, 64) | 152.97 | 304.37 | 123.38 |  87.39 |    -42% |    -19% |
| scatter_max / indices=10000, dst_shape=(1000, 64), src_shape=(10000, 64) |  14.71 |   3.86 |  56.15 |   1.67 |    -88% |   +281% |

We now see an overall 42-88% slower runtime compared to CPU runs with torch. Also, runtimes seem to be the same for `scatter_add` and `scatter_max`.

**Comment by awni** [comment]
> However, this approach isn't practical in real-world scenarios

Can you explain what you mean? I thought most mlx-graph use cases are similarly iterative numerical computations (just like SGD etc with standard neural nets)

In those computations the first iteration is almost always much more costly since you have to load a bunch of kernels to memory (possibly some compilation as well). That seems quite reasonable to expect  (and ignore) as usually there are enough iterations to amortize the startup cost. In almost all the benchmarks we do, we have a warmup so that we aren't measuring this cost.

**Comment by TristanBilot** [comment]
Ok I see, sorry for the confusion. You are right, this is an iterative process like in standard NNs. I wasn't sure about the reason of this first call overhead, but as you explained this should be amortized once the program is running. So we can keep the first benchmark as baseline I guess.

**Comment by TristanBilot** [comment]
In [this article](https://www.pyg.org/ns-newsarticle-accelerating-pyg-on-intel-cpus) is explained how `torch_scatter` improves speedup of scattering operations on CPU architecture.

**Comment by TristanBilot** [comment]
This [cuda kernel](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/cuda/scatter_cuda.cu) of `torch_scatter` may be of interest.

**Comment by awni** [comment]
I think atomics are way faster in CUDA than METAL 🤔 but their CUDA scatter implementation does look like it is using atomics, so not that different from ours.

I don't know of a good way to get rid of atomics without sorting the indices before applying the scatter. In that case you could so something like:

1. Sort indices
2. Threads / threadgroups over indices
3. Atomic write the resulting value(s) to the output

It's not trivial to implement though. And requires a fast sort which we likely need to work on. There may be a simpler solution I am not certain.

**Comment by francescofarina** [comment]
@awni when you say it requires a fast sort, do you already have a direction in mind? Perhaps a radix sort for this use case?

**Comment by TristanBilot** [comment]
I'm not sure to get why we want to sort indices here, is it to group same indices in same thread groups?

**Comment by TristanBilot** [comment]
Okay I understand well now. I see that the Metal sort implementation is inspired by cub, right?
I'll try to do a first benchmark to see whether the current implementation of sort could be used.

**Comment by TristanBilot** [comment]
It seems that `sort()` on GPU cannot handle arrays with >2M elements. 
`ERROR: test case THREW exception: [sort] GPU sort cannot handle sort axis of >= 2M elements, got array with sort axis size 10000000. Please place this operation on the CPU instead.`

**Comment by awni** [comment]
I'm guessing the limit is due to INT_MAX. Do you need larger than 2M sorts? That's pretty big 😄 , I see you have some benchmarks with indices up to 1M, is that practical? Are bigger index arrays common?

**Comment by TristanBilot** [comment]
Ahha, unfortunately yes graphs can easily reach more than 2 million edges so we have to consider larger arrays of indices. One advantage of unifier memory is also to run locally large computations with many indices.

Anyways, we can fix that once we know if it's worth to use a sort-based method for scattering.
Here are some new benchmarks of `scatter` compared to `sort` (CPU/GPU):

```python
# Scattering

# 1k indices
Timing (gpu source=100 indices=1000) lambda_scatter ... 0.40174 msec
Timing (gpu source=1000 indices=1000) lambda_scatter ... 0.33038 msec
Timing (gpu source=10000 indices=1000) lambda_scatter ... 0.34188 msec

# 100k indices
Timing (gpu source=100 indices=100000) lambda_scatter ... 5.7112 msec
Timing (gpu source=1000 indices=100000) lambda_scatter ... 5.6597 msec
Timing (gpu source=10000 indices=100000) lambda_scatter ... 5.6918 msec

# 1M indices
Timing (gpu source=100 indices=1000000) lambda_scatter ... 61.252 msec
Timing (gpu source=1000 indices=1000000) lambda_scatter ... 61.215 msec
Timing (gpu source=10000 indices=1000000) lambda_scatter ... 61.228 msec

# Sorting

# 1k indices
Timing (gpu indices=1000) lambda_sort ... 0.26064 msec

# 100k indices
Timing (gpu indices=100000) lambda_sort ... 0.59282 msec

# 1M indices
Timing (gpu indices=1000000) lambda_sort ... 1.2972 msec
```

Even with 1M indices, `sort` is really efficient (1.2972 msec is nuts compared to 61 msec), so it might be interesting to use a sort-based kernel for large arrays of indices, WDYT?

**Comment by awni** [comment]
> Even with 1M indices, sort is really efficient (1.2972 msec is nuts compared to 61 msec), so it might be interesting to use a sort-based kernel for large arrays of indices, WDYT?

Yea that's fast enough to merit trying a sort-based scatter and I think it could be made even faster if needed.

**Comment by awni** [comment]
> Ahha, unfortunately yes graphs can easily reach more than 2 million edges so we have to consider larger arrays of indices. One cool thing about unified memory is also that we can run locally large computations with many indices.

Makes sense. I haven't looked at our sort in a while, but we can probably do `uint` trivially (so double the limit from 2 to 4 million), 64-bit may require bigger changed.

**Comment by TristanBilot** [comment]
Great, can you tell me if my first intuition regarding the sort-based kernel is right:
* indices are sorted with radix sort, along with the corresponding values
* each thread receives a fixed-size slice of the sorted indices with associated values and computes a local scatter using a for loop (maybe an unrolled loop?)
* the local scatter of each index within the thread is written to the output array via an atomic

**Comment by vj-krish** [comment]
@TristanBilot @awni should we close this now?

**Comment by TristanBilot** [comment]
I think we can close! If there are specific cases to improve in the future, we'll open another issue :)
